### ISSUE TITLE
How to increase and decrease number of threads in thread pool

### ISSUE THREAD

**Issue by tanyifeng** [criticalstack/libevhtp#100]
# Details
I'm trying to manage threads pool according to number of http requests. Intend to assign a thread for every reqeust, and the thread can exit when it finished handle the request. 

# Steps or code to reproduce the problem.
Using current libevhtp version, we could hang in such case:
## Use EVTHR_SHARED_PIPE mode.
When the number of concurrent requests is more than the thread numbers we set, we would block the rest reqeust if no callback in all thread is returned.

## Not use EVTHR_SHARED_PIPE mode.
According to  code [thread.c](https://github.com/criticalstack/libevhtp/blob/6cb403a74866c4cfd0b8fbe00e02851d3019a9fa/thread.c#L349), we always distribute reqeust n % num_threads to the same thread, if a callback is blocked, the following reqeust will all be blocked.

# Version
Latest version

Can you give me some advice for such purpose. Since [event_base_loop](https://github.com/criticalstack/libevhtp/blob/6cb403a74866c4cfd0b8fbe00e02851d3019a9fa/thread.c#L121) will not exit if it has finished one request, and also I can not find a way to check if event base is finished one request or not.

**Comment by NathanFrench** [comment]
Are you checking the return of `evthr_pool_defer`? If you are getting blocked, it may be due to a return of `EVTHR_RES_RETRY` (it's like `EAGAIN` for threads).

**Comment by tanyifeng** [comment]
I mean the `evthr_pool_defer` is success, but if there are no enough threads to handle all the requests, one thread may have two or more callback waiting to be called, if the previous callback is hang, all the remaining callback will not be called. So I want to assign a thread for every reqeust, the hanged callback will have no effect to others.

There is a simple example code, base on branch master.
[test_hang_patch.txt](https://github.com/criticalstack/libevhtp/files/2174388/test_hang_patch.txt)

1. Run test_test with 2 threads.
 `test_test -t -n 2`, 

2. Send 5 concurrent requests.

3. Here is the results.
![image](https://user-images.githubusercontent.com/28038581/42427116-57c99752-835f-11e8-8c84-2e781f1882a5.png)

The server received 5 requests and writed success by `evthr_pool_defer`, the request 3, 4, 5 handled after callback 1,2 is returned after 10 seconds. If callback 1,2 never returned, reqeust 3,4,5 will also never be handled. If we can create 5 threads for 5 requests even if we specified `-n 2` when start server, then recycle 3 threads when callback is returned, ensure all requests get handled.

**Comment by NathanFrench** [comment]
If you have a long-running function in a thread, of course, it will stall the event loop for that thread.

Correct me if I'm wrong: what you are looking for is a way to have threads automatically start and stop based on the contention of currently running threads, correct? Kind of how apache and nginx do it.

You also have the option of rolling your own pool, `evthr_pool_t` is just a wrapper around multiple `evthr_t`'s:

```C
evthr_t      * evthr_new(evthr_init_cb, void *);
int            evthr_start(evthr_t * evthr);
evthr_res      evthr_stop(evthr_t * evthr);
evthr_res      evthr_defer(evthr_t * evthr, evthr_cb cb, void *);
```

**Comment by tanyifeng** [comment]
Yes, that's what I want to say. I'm trying to manage my own pool, but can not find a way to check if event base is finished one request or not, if we want to recycle a thread, must have a method to check it.

**Comment by NathanFrench** [comment]
Auto-scaling threads sound like a great idea. I will start working on it in the morning.

**Comment by tanyifeng** [comment]
Cool. Hope to know your solution.
